### PR TITLE
feature/1809 allow for custom error logic for self service request hook

### PR DESCRIPTION
**Pull request by SEQUOIIA**
For context, see: https://github.com/dfds/cloudplatform/issues/1809

- Added generalised Error hook & helper classes
- Make sure components spawned from .map has a key so React can identify components
- Updated SelfServiceApi hook to use the new Error hook
- Added error handler for useCapabilityById (mostly for demonstrating how a custom handler can be attached)

---

~~Not sure if I'm happy with the naming of "_NewErrorTriggerRequestBuilder_" 😅~~

Would appreciate feedback of any kind. Perhaps it's too convoluted?

### CODE REVIEW

File: src/src/hooks/Error.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+                      showError(params.msg);
+                      return;
+                  }
+
+                  showError(options.msg);
+              } else {
+                  showError(options.msg);
+              }
+          }
+      }
```
Comment (kralle333) [inline src/src/hooks/Error.js:40]: It is quite difficult to read the flow here, could you invert some of the statements (+ return) to reduce nesting and then maybe also adding a comment that explain what the different scenario we can encounter here.

File: src/src/hooks/SelfServiceApi.js [hunk 2]
```diff
@@ -34,31 +36,33 @@ export function useSelfServiceRequest() {
...
+                triggerError(NewErrorTriggerRequestBuilder()
+                  .setHttpResponse(httpResponse)
+                  // .setHandler((params) => {
+                  //   if (params.httpResponse.headers.get("Content-Type") === "application/problem+json") {
+                  //       const { detail } = await httpResponse.json();
+                  //       setErrorMessage(detail);
+                  //   } else {
+                  //       setErrorMessage("Oh no! We had an issue while retrieving data from the api. Please reload the page.");
+                  //   }
+                  // })
```
Comment (kralle333) [inline src/src/hooks/SelfServiceApi.js:48]: should we delete these lines?

File: src/src/hooks/SelfServiceApi.js [hunk 3]
```diff
@@ -34,31 +36,33 @@ export function useSelfServiceRequest() {
...
+                  .build());
+
             }
         } catch (error) {
-            setErrorMessage(error.message);
+            triggerError(NewErrorTriggerRequestBuilder()
```
Comment (kralle333) [inline src/src/hooks/SelfServiceApi.js:53]: You were asking for naming right? Could we name it NewTriggerErrorRequestBuilder to keep the same word ordering (triggerError vs errorTrigger).

Do we need the word Request in the name? Or can we omit it or find a better word.

File: src/src/ErrorDisplay.js [hunk 4]
```diff
@@ -4,14 +4,14 @@ import Toast from 'components/Toast/index.js'
...
-  
+
   if (error.length === 0 ) {
     return null;
   }
 
   return (
     <div>
-        {(error.map(e => <Toast message={e}></Toast>))}   
+        {(error.map((e, index) => <Toast key={index} message={e}></Toast>))}
```
Comment (kralle333) [inline src/src/ErrorDisplay.js:14]: What does the `key={index}` thing do :)

File: src/src/misc/error.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+  setHandler(input) {
+    this.val.handler = input;
+    return this;
+  }
+
+  build() {
+    return this.val;
+  }
+}
```
Comment (kralle333) [inline src/src/misc/error.js:40]: Not sure if using null like this is idiomatic javascript, but it surely will make code harder to write because you have to have null checks everywhere. Is there a way to rewrite this so that we can get rid of null values and checks?

**Review by kralle333 — CHANGES_REQUESTED** [review]
only small request for changes :) will be very nice to get in

File: src/src/hooks/SelfServiceApi.js (hunk 2 above)
Comment (SEQUOIIA) [reply src/src/hooks/SelfServiceApi.js]: Removed in https://github.com/dfds/selfservice-portal/pull/25/commits/baed7bb34952f2db0f9fa11fd79886a7ba915e48

File: src/src/hooks/Error.js (hunk 1 above)
Comment (SEQUOIIA) [reply src/src/hooks/Error.js]: I... made an attempt I think - https://github.com/dfds/selfservice-portal/pull/25/commits/f4f0f43d951a8055836c6b8d8286e7fe1b80f75d 😅

File: src/src/hooks/SelfServiceApi.js (hunk 3 above)
Comment (SEQUOIIA) [reply src/src/hooks/SelfServiceApi.js]: Thought about it for a bit, updated it to be ErrorContext instead of ErrorTriggerRequest (https://github.com/dfds/selfservice-portal/pull/25/commits/62b93bd52ae0472919825691ab39f757d92bb898). Thoughts?

File: src/src/ErrorDisplay.js (hunk 4 above)
Comment (SEQUOIIA) [reply src/src/ErrorDisplay.js:14]: https://react.dev/learn/rendering-lists#why-does-react-need-keys

TLDR: when rendering an array of components, keys allow React to differentiate between the components

Unfortunately the e variable in this case in the actual error message, and there can occasionally be duplicates, so I instead opted for using the index value that tells us the location of e within the array that is generated by .map

File: src/src/misc/error.js (hunk 5 above)
Comment (SEQUOIIA) [reply src/src/misc/error.js]: Agree that it's not ideal, but can't currently think of a nicer alternative. 

If all the values that is currently assigned to null is given a default value, I'd instead have to check all of them in a different way. If httpResponse has a default value, I need to make sure that the object contains the keys I'd expect from a properly filled out httpResponse before using it. If handler is given a default value, I can't know for sure whether the handler has been replaced by the user, thus giving it higher priority in the error handling flow. Well, I could build a different way of determining whether it's been switched out or not, but null also works and is much less effort.

Given the nature of how ErrorContext is used, there will almost always be some values that shouldn't be set. If a network error happens, then _httpResponse_ should never be populated, instead _error_ would be. In most cases, if _httpResponse_ is populated, then _error_ rarely is. _msg_  and _handler_ can be populated regardless of any of these, that would be up to the user to set that one in most cases. _showError_ is probably the only one that you can almost always rely on being available, since it is set by the useError Hook.

File: src/src/hooks/SelfServiceApi.js (hunk 3 above)
Comment (kralle333) [reply src/src/hooks/SelfServiceApi.js]: sounds good

File: src/src/hooks/Error.js (hunk 1 above)
Comment (kralle333) [reply src/src/hooks/Error.js]: wow extremely nice!

File: src/src/misc/error.js (hunk 5 above)
Comment (kralle333) [reply src/src/misc/error.js]: Wow a whole novel :D I was just curious, don't read too much into my comment.

File: src/src/ErrorDisplay.js (hunk 4 above)
Comment (kralle333) [reply src/src/ErrorDisplay.js:14]: ok cool, thanks for the link

**Review by kralle333 — APPROVED** [review]

**Comment by kralle333** [comment]
Very good job. Thanks for listening to my comments and being so thorough!!!
